libraries/AP_GPS/AP_GPS_VecNAV.cpp

```diff
@@ -0,0 +1,345 @@
+      packetsize = 0;
+      packetsize++;
+      msgState = 1;
+      //hal.console->printf("Received header\n");
+      //printf("Received header\n");
+    }
+    break;
+
+  case 1:
+    // Groups active    
+    dataPl[packetsize-1] = c;
+    packetsize++;
+    msgState = 2;
+    count = 0;
+    //printf("Group %02x is active\n",c);
+    //hal.console->printf("Group %02x is active\n",c);
+    
+    break;
+    
+  case 2:
+    // Active fields in each group
+    dataPl[packetsize-1] = c;
+    packetsize++;
+    if(count == 0){
+      fieldSize1 = 0;
+      fieldSize1 = fieldSize1 | c;
+      count++;
+    }
+    else if(count == 1){
+      fieldSize1 = (fieldSize1 << 8) | c;
+      count++;
+    }
+    else if(count == 2){
+      fieldSize2 = 0;
+      fieldSize2 = fieldSize2 | c;
+      count++;
+    }
+    else if(count == 3){
+      fieldSize2 = (fieldSize2 << 8) | c;      
+      count++;
+    }
+    else if(count == 4){
+      fieldSize3 = 0;
+      fieldSize3 = fieldSize3 | c;
+      count++;
+    }
+    else if(count == 5){
+      fieldSize3 = (fieldSize3 << 8) | c;      
+      count = 0;
+      msgState = 3;
+    }
+
+    break;
+
+  case 3:
+    //Payload
+    //printf("datasize = %d\n",packetsize);
+    dataPl[packetsize-1] = c;
+    packetsize++;
+    count++;    
+    if(count == (DATA_LEN - HEADER - CRC_LEN)){      
+      //printf("Receive payload, count=%d\n",count);
+       //hal.console->printf("Receive payload, count=%d\n",count);
+      
+      msgState = 4;
+      count = 0;      
+    }
+
+    break;
+    
+  case 4:
+    //CRC
+    //printf("datasize = %d\n",packetsize);
+    dataPl[packetsize-1] = c;
+    packetsize++;
+    if(count == 0){
+      CRC = 0;
+      CRC = CRC | c;
+      count++;
+    }
+    else if(count == 1){
+      CRC = (CRC << 8) | c;
+      msgState = 5;
+      count = 0;
+      
+      //printf("Received checksum\n");
+    }
+
+    break;
+
+  case 5:
+    // Validate data    
+    CRCV = calculateCRC(dataPl,DATA_LEN);
+    msgState = 0;
+    count = 0;
+    if(CRCV == 0x0000){
+      //printf("Valid data obtained\n");
+      //hal.console->print("Valid data obtained");
+
+      p = dataPl + HEADER;
+      ExtractDataGroup1(p,msgG1);
+      
+      p = dataPl + HEADER + GROUP1_LEN;
+      ExtractDataGroup4(p,msgG4);
+
+      p = dataPl + HEADER + GROUP1_LEN + GROUP4_LEN;
+      ExtractDataGroup6(p,msgG6);
+      
+      return true;
+    }
+    else{      
+      return false;
+    }
+    
+    break;
+  }//end of switch
+
+  return false;
+}
+
+bool AP_GPS_VecNAV::read(void){
+
+    int16_t numc;
+    bool parsed = false;
+      
+    numc = port->available();
+
+    //hal.console->printf("Available data %d\n",numc);
+
+    while (numc--) {
+        
+        char c = port->read();
+        //hal.console->printf("%c",c);
+        parsed = ProcessGPSMessage(c,&msg1,&msg4,&msg6);
+
+        if(parsed){
+            
+            state.location.lng    = (int32_t) (msg6.PosLla[1]*1e7);
```

Comment:
My reading of the VN-200 data sheet says that this is the INS position, rather then the GPS position. This means its already been filtered once by the VN. This message should be built with the msg4 information instead to avoid feeding the EKF prefiltered data.